references/cli/delete.go

```diff
@@ -17,79 +17,384 @@ limitations under the License.
 			return err
 		}
-		for _, app := range args {
-			o.AppName = app
-			userInput := NewUserInput()
-			if !assumeYes {
-				userConfirmation := userInput.AskBool(fmt.Sprintf("Do you want to delete the application %s from namespace %s", o.AppName, o.Namespace), &UserInputOptions{assumeYes})
-				if !userConfirmation {
-					return fmt.Errorf("stopping Deleting")
+	}
+
+	_, _ = fmt.Fprintf(cmd.OutOrStdout(), "Delete appplication %s/%s succeeded\n", app.Namespace, app.Name)
+	return nil
+}
+
+func (opt *DeleteOptions) orphan(ctx context.Context, f velacmd.Factory, app *v1beta1.Application) error {
+	if !slices.Contains(app.GetFinalizers(), oam.FinalizerOrphanResource) {
+		meta.AddFinalizer(app, oam.FinalizerOrphanResource)
+		if err := f.Client().Update(ctx, app); err != nil {
+			return fmt.Errorf("failed to set orphan resource finalizer to application %s/%s: %w", app.Namespace, app.Name, err)
+		}
+	}
+	return nil
+}
+
+func (opt *DeleteOptions) forceDelete(ctx context.Context, f velacmd.Factory, app *v1beta1.Application) error {
+	return wait.PollImmediate(3*time.Second, 1*time.Minute, func() (done bool, err error) {
+		err = f.Client().Get(ctx, client.ObjectKeyFromObject(app), app)
+		if kerrors.IsNotFound(err) {
+			return true, nil
+		}
+		rk, err := resourcekeeper.NewResourceKeeper(ctx, f.Client(), app)
+		if err != nil {
+			return false, fmt.Errorf("failed to create resource keeper to run garbage collection: %w", err)
+		}
+		if done, _, err = rk.GarbageCollect(ctx); err != nil && !kerrors.IsConflict(err) {
+			return false, fmt.Errorf("failed to run garbage collect: %w", err)
+		}
+		if done {
+			meta.RemoveFinalizer(app, oam.FinalizerResourceTracker)
+			meta.RemoveFinalizer(app, oam.FinalizerOrphanResource)
+			if err = f.Client().Update(ctx, app); err != nil && !kerrors.IsConflict(err) && !kerrors.IsNotFound(err) {
+				return false, fmt.Errorf("failed to update app finalizer: %w", err)
+			}
+		}
+		return false, nil
+	})
+}
+
+func (opt *DeleteOptions) deleteResource(ctx context.Context, f velacmd.Factory, mr v1beta1.ManagedResource, app *v1beta1.Application) error {
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(mr.GroupVersionKind())
+	if err := f.Client().Get(multicluster.WithCluster(ctx, mr.Cluster), mr.NamespacedName(), obj); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	if !resourcekeeper.IsResourceManagedByApplication(obj, app) {
+		return nil
+	}
+	return resourcekeeper.DeleteManagedResourceInApplication(ctx, f.Client(), mr, obj, app)
+}
+
+func _getManagedResourceSource(mr v1beta1.ManagedResource) string {
+	src := "in cluster local"
+	if mr.Cluster != "" {
+		src = fmt.Sprintf("in cluster %s", mr.Cluster)
+	}
+	if mr.Namespace != "" {
+		src += fmt.Sprintf(", namespace %s", mr.Namespace)
+	}
+	groups := strings.Split(mr.APIVersion, "/")
+	group := "." + groups[0]
+	if len(groups) == 0 {
+		group = ""
+	}
+	return fmt.Sprintf("%s%s %s %s", strings.ToLower(mr.Kind), group, mr.Name, src)
```

Comment:
The option is not clear. If there are multiple resource trackers that include the same resource, how do the users select the resource that wanted to delete?

Could we mark the error resource?

Reply:
That resource will be removed from all ResourceTrackers.

I think marking error resource itself would not solve the probelm?